### PR TITLE
Automatically create the hash data directory when using an alternative

### PR DESCRIPTION
**Pull request by adreed-msft**
Under current behavior, AzCopy appears to expect a similar file structure when specifying an alternate data directory for hashes. To remediate this issue (#2584), automatically create the parent directories when saving hash data to the disk.

### CODE REVIEW

File: common/hash_data_adapter_hidden_files.go [hunk 1]
```diff
@@ -6,15 +6,18 @@ import (
...
 	"os"
 	"path/filepath"
+	"sync"
 )
 
 type HiddenFileDataAdapter struct {
 	hashBasePath string // "" == dataBasePath
 	dataBasePath string
+
+	createHashDirOnce *sync.Once
```
Comment (gapra-msft) [inline common/hash_data_adapter_hidden_files.go:16]: Should we use this in the lines added below to create the hash dir? I am missing where this is used currently.

**Review by gapra-msft — COMMENTED** [review]
Test changes look good. I have an open question about the sync.Once

File: common/hash_data_adapter_hidden_files.go (hunk 1 above)
Comment (adreed-msft) [reply common/hash_data_adapter_hidden_files.go:16]: It was a remnant of prior code changes in this PR-- I forgot to remove this, but, I'll trim it now.

Long story short-- We can't know the entire directory structure up front here, so we have to create it as we go.

**Review by gapra-msft — APPROVED** [review]

**Review by siminsavani-msft — APPROVED** [review]

File: e2etest/newe2e_sas_generation.go [hunk 2]
```diff
@@ -205,6 +206,8 @@ func (vals GenericAccountSignatureValues) AsBlob() BlobSignatureValues {
 func (vals GenericAccountSignatureValues) AsFile() FileSignatureValues {
 	s := vals.withDefaults()
 
+	s.Permissions = strings.ReplaceAll(s.Permissions, "a", "") // remove 'a', because it's invalid and causes panics.
```
[the remark this answers is not in the document]
Comment (adreed-msft) [reply e2etest/newe2e_sas_generation.go:209]: Yes, it is not valid on a files signature.

**Review by gapra-msft — APPROVED** [review]

**Pull request merged**